assets/js/cell/live_editor/monaco_editor_adapter.js

```diff
@@ -32,19 +36,38 @@ export default class MonacoEditorAdapter {
    * Applies the given delta to the editor content.
    */
   applyDelta(delta) {
+    const isStandaloneChange = delta.ops.some((op) => {
+      if (isDelete(op)) {
+        return true;
+      }
+
+      if (isInsert(op)) {
+        return op.insert.match(/\s+/);
+      }
+
+      return false;
+    });
+
+    // Explicitly close the last stack element when the remote
+    // change inserts whitespace or deletes text. Otherwise
+    // merge subsequent remote changes whenever possible.
+    if (isStandaloneChange || !this.isLastChangeRemote) {
+      this.editor.getModel().pushStackElement();
+    } else {
+      this.editor.getModel().popStackElement();
```

Comment:
Note: the API naming is slightly misleading here, the push/pop methods only really open/close the last stack element, they don't actually modify the stack itself.